### ISSUE TITLE
"Too many calls to releaseReadLock " assertion failure opening corrupt database

### ISSUE THREAD

**Issue by mz2** [couchbase/couchbase-lite-ios#977]
I suppose this could be seen as an FMDatabase level bug, or not a bug at all, but CBLDatabase -open: will not treat attempting to open a corrupt database as an error case, but an exception is thrown (inside FMDatabase). If this case is not treated by either of these layers, as a client to CBL one needs to add a @try / @catch around CBLDatabase -open: to deal with real world data, which would be unfortunate.

Please bear in mind also that the below stack frames are for a rather old (~1.0 era) CBL so this may no longer be relevant, at which case please close. 

```
> 11/11/15 20:22:56,639 Manuscripts[642]: Unknown error calling sqlite3_step 
> (11: database disk image is malformed) rs 
> 11/11/15 20:22:56,639 Manuscripts[642]: *** Assertion failure in 
> -[CBL_FMDatabase releaseReadLock], 
> /Users/mz2/Projects/Manuscripts/External/Feather/External/couchbase-lite-ios/vendor/fmdb/src/FMDatabase.m:231 
> 11/11/15 20:22:56,654 Manuscripts[642]: An uncaught exception was raised 
> 11/11/15 20:22:56,655 Manuscripts[642]: Too many calls to releaseReadLock 
> 11/11/15 20:22:56,655 Manuscripts[642]: ( 
> 0 CoreFoundation 0x00007fff97b10bd2 
> __exceptionPreprocess + 178 
> 1 libobjc.A.dylib 0x00007fff9e9a2dd4 
> objc_exception_throw + 48 
> 2 CoreFoundation 0x00007fff97b158aa +[NSException 
> raise:format:arguments:] + 106 
> 3 Foundation 0x00007fff91ea7f46 
> -[NSAssertionHandler 
> handleFailureInMethod:object:file:lineNumber:description:] + 198 
> 4 CouchbaseLite 0x000000010c9d2b8e 
> -[CBL_FMDatabase releaseReadLock] + 141 
> 5 CouchbaseLite 0x000000010c9d3aef 
> -[CBL_FMDatabase executeQuery:withArgumentsInArray:orVAList:] + 1269 
> 6 CouchbaseLite 0x000000010c9d4ab5 
> -[CBL_FMDatabase(FMDatabaseAdditions) intForQuery:] + 160 
> 7 CouchbaseLite 0x000000010c97b14b 
> -[CBLDatabase(Internal) open:] + 106 
> 8 CouchbaseLite 0x000000010c9b93c8 -[CBLManager 
> databaseNamed:error:] + 62 
```

**Comment by snej** [comment]
That's a bug in [our modified version of] FMDatabase, not intentional behavior. I'm pretty sure this is something we've fixed since 1.0; I remember at least one bug-fix involving `releaseReadLock`. (You can read through the git history of FMDatabase.m to check.)

> as a client to CBL one needs to add a @try / @catch around CBLDatabase -open:

Please don't do that, unless your `@catch` block immediately quits the app. By the time you've caught the exception, CBL will be in an "undefined state" since a bunch of code got aborted halfway through, and there's no telling what will happen the next time you call it.

**Comment by snej** [comment]
This was fixed by cdd2d90.
